Update draft-lcurley-warp.md (#1)
wilaw/warp-draft: draft-lcurley-warp.md

                             └───────┘                                                        
~~~~

## MediaSession

MediaSession represents a highest level abstraction or a container from 
a media application perspective using WARP as the media delivery protocol. 

> A MediaSession has a defined start and contains one or more MediaStreams produced by a publisher over time.

Author: ```suggestion
A MediaSession has a defined start and contains one or more MediaStreams produced by a publisher and optionally transferred by relays and consumed by subscribers over time. 
```

A MediaSession may represent a Broadcast for a live media 
session, an Interactive conference meeting session, a tile-based VR delivery or another media application. 

A MediaSession is described by an alphanumeric MediaSessionID. This mediaSessionID may be a URL, a number, or a combination of the two. A MediaSessionID MUST be globally unique when produced by a publisher. Relays MAY modify the MediaSessionID as objects transit their network, but the original MediaSession ID must be restored when the objects egress their newtwork. 

`
Example valid mediaSessionIDs:
  acme.tv/broadcasts/channel8/alice
  987239847238792871
  5fcf18f9-5f19-4f82-93e7-a4bf21e0bfa6
`

## MediaStream

MediaStreams are the children of MediaSessions. A MediaStream contains a sequential series of MediaGroups.

MediaStreams hold data of a consistent MimeType. Examples of MediaStreams would be 1080p video of a game, or AAC audio, or captions. 

MediaStreams are identified by an alphanumberic MediaStreamID and are independently subscrible assets within a MediaSession.

Examples of valid MediaStreamIDs:

`
  /hd/
  23
  desktopScreenShare
`

> Note - MediaStreams should not be confused with QUIC streams although certain 
> applications might allow mapping of MediaStreams to one or more QUIC Streams.		

Author: ```suggestion

```

time to live, and other information aiding the caching/forwarding decision at the Relays. 

Each MediaObject is identified by a sequentially increasing integer, called MediaObjectId, starting at 0. 

Media Objects represent the cacheable entity within the Warp architecture. For the purposes of publishing, 
caching and retrieval, a Media Object is fully identified by the combination of the following identifiers:

`
Object cache key :=   MediaSessionID | MediaStreamId | MediaGroupId | MediaObjectId
`
These identitiers are stored as separate fields in the header of each MediaObject, to facilitate unambiguous extraction by each relay. 

## Scope / Goals of the Data Model

### Subscribers

One of the goals of the proposed data model is to allow subscribers, 
within a MediaSession, to request the right granularity of the media 
resources in a way that is independent of application and a 
given transport mapping.

To that extent, the proposed data model enables following ways for the subscribers 
(end-points or Relays) to ask for the data:

>     -  Request a specific MediaStream. Such a request enables subscribers to 
>        receive all future media groups and objects under a 
>        given MediaStream.

Author: ```suggestion
    - Request an entire mediaSession by providing only the MediaSessionID in the subscribe request and omitting all other IDs.
    -  Request a specific MediaStream. Such a request enables subscribers to 
       receive all future media groups and objects under a 
       given MediaStream.
```

>     -  Request a specific MediaStream. Such a request enables subscribers to 
>        receive all future media groups and objects under a 
>        given MediaStream.

Author: ```suggestion
    -  Request a specific MediaSession.  Such a request enables subscribers to 
       receive all future mediastreams,  mediagroups and mediaobjects under a 
       given MediaSession.
    -  Request a specific MediaStream. Such a request enables subscribers to 
       receive all future media groups and objects under a 
       given MediaStream.
```

    -  Request a specific MediaStream starting at a specific MediaGroup. Since the MediaGroup identification is ordinal, subsubscribers may request for example content starting from the last, or 3rd last, MediaGroup available in the cache, or from the first one for which MediaGroupID  > N
       

> The MediSession ID to given to each subscriber via an out-of-band mechanism. The subscriber uses this MediaSessionID to request the CATALOG, which in turn provides information to the client about the MediaStreamD and MediaGroup IDs which it can use to construct its subscription requests. 

Author: ```suggestion
The MediSession ID to given to each subscriber via an out-of-band mechanism. The subscriber uses this MediaSessionID to request the CATALOG, which in turn provides information to the client about the MediaStreamIDs which it can use to construct its subscription requests. 
```

              `─────'                                           │        .─────.
                                                                └───────(  S2   )
     pub: acme.com/meetings/m123/alice/video                             `─────'
     pub: acme.com/meetings/m123/alice/audio
                                                        sub:acme.com/meetings/m123/alice/audio
                                                        sub:acme.com/meetings/m123/alice/video
                                                        sub:acme.com/meetings/m123/bob/audio
                                                        sub:acme.com/meetings/m123/bob/video

~~~~

The above picture shows Warp media delivery, where 
a tree topography is formed with multiple relays in the network. 
The example  has 4 participants with Alice and Bob

> being the publishers and S! and S2 being the subscribers. 

Author: ```suggestion
being the publishers and S1 and S2 being the subscribers. 
```
